crates/wysiwyg/src/composer_model/code_block.rs

```diff
@@ -435,149 +247,114 @@ where
+                        c, range, first_leaf, last_leaf,
+                    ));
                 }
-                DomNode::Container(container.clone_with_new_children(children))
+                // Then post-process the containers to add line breaks and move selection if needed
+                if container.is_block_node() || container.is_list_item() {
+                    if container.is_list() {
+                        // Add line break before the List if it's not at the start
+                        if node.handle().index_in_parent() > 0 {
+                            children.insert(0, DomNode::new_text("\n".into()));
+                            if is_before {
+                                self.state.start += 1;
+                            }
+                        }
+                    } else if container.is_list_item() {
+                        // Every list item adds a line break at the end
+                        // So "<ul><li>Item A</li><li>Item B</li></ul>End"
+                        // Will become: "<pre>Item A\nItem B\nEnd</pre>"
+                        children.push(DomNode::new_text("\n".into()));
+                        if is_before {
+                            self.state.start += 1;
+                        }
+                        if !is_after {
+                            self.state.end += 1;
+                        }
+                    } else if *container.kind() == ContainerNodeKind::Quote {
+                        // We add a new line where it's needed
+                        let added_at_start =
+                            if container.handle().index_in_parent() == 0 {
+                                children.push(DomNode::new_text("\n".into()));
+                                false
+                            } else {
+                                children
+                                    .insert(0, DomNode::new_text("\n".into()));
+                                true
+                            };
+                        if is_in_selection {
+                            let location = range
+                                .locations
+                                .iter()
+                                .find(|l| l.node_handle == container.handle())
+                                .unwrap();
+                            if location.starts_inside()
+                                && location.ends_inside()
+                            {
+                                // Do nothing
+                            } else if location.starts_inside() {
+                                if added_at_start {
+                                    self.state.start += 1;
+                                }
+                                self.state.end += 1;
+                            } else if location.ends_inside() {
+                                self.state.start += 1;
+                                self.state.end += 1;
+                            }
+                        }
```

Comment:
this definitely could be a separate function that handles the Quote case.

Reply:
Extracted to functions in 81750202bd4f1e5ebdfeebcefdd7eb4d9837fb87